File: src/esengine/platform/input/Input.cpp
/**
 * @file    Input.cpp
 * @brief   Input system implementation
 * @details Manages touch, mouse, and keyboard input state tracking.
 *
 * @author  ESEngine Team
 * @date    2025
 *
 * @copyright Copyright (c) 2025 ESEngine Team
 *            Licensed under the MIT License.
 */

#include "Input.hpp"
#include "../../core/Log.hpp"

namespace esengine {

// Static member definitions
std::array<TouchState, MAX_TOUCH_POINTS> Input::touchStates_;
std::array<TouchState, MAX_TOUCH_POINTS> Input::prevTouchStates_;
std::unordered_map<u32, bool> Input::keyStates_;
std::unordered_map<u32, bool> Input::prevKeyStates_;
glm::vec2 Input::mousePosition_{0.0f};
glm::vec2 Input::scrollDelta_{0.0f};

void Input::init() {
    for (auto& state : touchStates_) {
        state = TouchState{};
    }
    for (auto& state : prevTouchStates_) {
        state = TouchState{};
    }
    keyStates_.clear();
    prevKeyStates_.clear();
    mousePosition_ = glm::vec2(0.0f);
    scrollDelta_ = glm::vec2(0.0f);

    ES_LOG_INFO("Input system initialized");
}

void Input::shutdown() {
    ES_LOG_INFO("Input system shutdown");
}

void Input::update() {
    // Store previous frame's state
    prevTouchStates_ = touchStates_;
    prevKeyStates_ = keyStates_;

    // Clear per-frame values
    scrollDelta_ = glm::vec2(0.0f);
}

// Touch input
bool Input::isTouchDown(u32 index) {
    if (index >= MAX_TOUCH_POINTS) return false;
    return touchStates_[index].active;
}

bool Input::isTouchPressed(u32 index) {
    if (index >= MAX_TOUCH_POINTS) return false;
    return touchStates_[index].active && !prevTouchStates_[index].active;
}

bool Input::isTouchReleased(u32 index) {
    if (index >= MAX_TOUCH_POINTS) return false;
    return !touchStates_[index].active && prevTouchStates_[index].active;
}

glm::vec2 Input::getTouchPosition(u32 index) {
    if (index >= MAX_TOUCH_POINTS) return glm::vec2(0.0f);
    return glm::vec2(touchStates_[index].x, touchStates_[index].y);
}

glm::vec2 Input::getTouchDelta(u32 index) {
    if (index >= MAX_TOUCH_POINTS) return glm::vec2(0.0f);
    const auto& state = touchStates_[index];
    return glm::vec2(state.x - state.startX, state.y - state.startY);
}

u32 Input::getTouchCount() {
    u32 count = 0;
    for (const auto& state : touchStates_) {
        if (state.active) ++count;
    }
    return count;
}

// Keyboard input
bool Input::isKeyDown(KeyCode key) {
    auto it = keyStates_.find(static_cast<u32>(key));
    return it != keyStates_.end() && it->second;
}

bool Input::isKeyPressed(KeyCode key) {
    u32 keyCode = static_cast<u32>(key);
    bool current = keyStates_.count(keyCode) && keyStates_[keyCode];
    bool previous = prevKeyStates_.count(keyCode) && prevKeyStates_[keyCode];
    return current && !previous;
}

bool Input::isKeyReleased(KeyCode key) {
    u32 keyCode = static_cast<u32>(key);
    bool current = keyStates_.count(keyCode) && keyStates_[keyCode];
    bool previous = prevKeyStates_.count(keyCode) && prevKeyStates_[keyCode];
    return !current && previous;
}

// Mouse input
glm::vec2 Input::getMousePosition() {
    return mousePosition_;
}

bool Input::isMouseButtonDown(u32 button) {
    // Mouse button 0 is treated as touch 0
    if (button == 0) {
        return isTouchDown(0);
    }
    return false;
}

// Event handlers
void Input::onTouchEvent(TouchType type, const TouchPoint& point) {
    // Find or allocate touch slot
    u32 index = MAX_TOUCH_POINTS;

    // Find existing touch with this ID
    for (u32 i = 0; i < MAX_TOUCH_POINTS; ++i) {
        if (touchStates_[i].active && static_cast<i32>(i) == point.id) {
            index = i;
            break;
        }
    }

    // If new touch, find empty slot
    if (index == MAX_TOUCH_POINTS && type == TouchType::Begin) {
        for (u32 i = 0; i < MAX_TOUCH_POINTS; ++i) {
            if (!touchStates_[i].active) {
                index = i;
                break;
            }
        }
    }

    if (index >= MAX_TOUCH_POINTS) return;

    auto& state = touchStates_[index];

    switch (type) {
    case TouchType::Begin:
        state.active = true;
        state.x = point.x;
        state.y = point.y;
        state.startX = point.x;
        state.startY = point.y;
        break;

    case TouchType::Move:
        state.x = point.x;
        state.y = point.y;
        break;

    case TouchType::End:
    case TouchType::Cancel:
        state.active = false;
        break;
    }

    // Also update mouse position for debugging
    if (index == 0) {
        mousePosition_ = glm::vec2(point.x, point.y);
    }
}

void Input::onKeyEvent(KeyCode key, bool pressed) {
    keyStates_[static_cast<u32>(key)] = pressed;
}

void Input::onScrollEvent(f32 deltaX, f32 deltaY) {
    scrollDelta_.x += deltaX;
    scrollDelta_.y += deltaY;
}

glm::vec2 Input::getScrollDelta() {
    return scrollDelta_;
}

}  // namespace esengine

File: sdk/src/input/Input.ts
/**
 * ESEngine TypeScript SDK - Input Management
 *
 * Provides unified input handling for touch and keyboard.
 */

import { Vec2, KeyCode, TouchType, TouchPoint, vec2 } from '../core/Types';

// Touch state
interface TouchState {
  id: number;
  position: Vec2;
  startPosition: Vec2;
  isDown: boolean;
  justPressed: boolean;
  justReleased: boolean;
}

// Key state
interface KeyState {
  isDown: boolean;
  justPressed: boolean;
  justReleased: boolean;
}

/**
 * Input - Static class for input management.
 */
export class Input {
  private static touches: Map<number, TouchState> = new Map();
  private static keys: Map<KeyCode, KeyState> = new Map();
  private static mousePosition: Vec2 = { x: 0, y: 0 };
  private static mouseButtons: Map<number, KeyState> = new Map();
  private static initialized: boolean = false;

  /**
   * Initialize the input system.
   */
  static init(): void {
    if (this.initialized) return;
    this.touches.clear();
    this.keys.clear();
    this.mouseButtons.clear();
    this.initialized = true;
  }

  /**
   * Shut down the input system.
   */
  static shutdown(): void {
    this.touches.clear();
    this.keys.clear();
    this.mouseButtons.clear();
    this.initialized = false;
  }

  /**
   * Update input state. Call once per frame after processing events.
   */
  static update(): void {
    // Reset "just" states for touches
    for (const touch of this.touches.values()) {
      touch.justPressed = false;
      touch.justReleased = false;
    }

    // Remove released touches
    for (const [id, touch] of this.touches) {
      if (!touch.isDown && !touch.justReleased) {
        this.touches.delete(id);
      }
    }

    // Reset "just" states for keys
    for (const key of this.keys.values()) {
      key.justPressed = false;
      key.justReleased = false;
    }

    // Reset "just" states for mouse buttons
    for (const button of this.mouseButtons.values()) {
      button.justPressed = false;
      button.justReleased = false;
    }
  }

  // ========== Touch Input ==========

  /**
   * Handle touch event from platform.
   */
  static onTouchEvent(type: TouchType, point: TouchPoint): void {
    let touch = this.touches.get(point.id);

    switch (type) {
      case TouchType.Begin:
        touch = {
          id: point.id,
          position: { x: point.x, y: point.y },
          startPosition: { x: point.x, y: point.y },
          isDown: true,
          justPressed: true,
          justReleased: false,
        };
        this.touches.set(point.id, touch);
        break;

      case TouchType.Move:
        if (touch) {
          touch.position = { x: point.x, y: point.y };
        }
        break;

      case TouchType.End:
      case TouchType.Cancel:
        if (touch) {
          touch.position = { x: point.x, y: point.y };
          touch.isDown = false;
          touch.justReleased = true;
        }
        break;
    }
  }

  /**
   * Check if a touch is currently down.
   */
  static isTouchDown(index: number = 0): boolean {
    const touch = this.getTouchByIndex(index);
    return touch !== undefined && touch.isDown;
  }

  /**
   * Check if a touch just started this frame.
   */
  static isTouchPressed(index: number = 0): boolean {
    const touch = this.getTouchByIndex(index);
    return touch !== undefined && touch.justPressed;
  }

  /**
   * Check if a touch just ended this frame.
   */
  static isTouchReleased(index: number = 0): boolean {
    const touch = this.getTouchByIndex(index);
    return touch !== undefined && touch.justReleased;
  }

  /**
   * Get the current touch position.
   */
  static getTouchPosition(index: number = 0): Vec2 {
    const touch = this.getTouchByIndex(index);
    return touch ? { ...touch.position } : { x: 0, y: 0 };
  }

  /**
   * Get the touch delta from start position.
   */
  static getTouchDelta(index: number = 0): Vec2 {
    const touch = this.getTouchByIndex(index);
    if (!touch) return { x: 0, y: 0 };
    return {
      x: touch.position.x - touch.startPosition.x,
      y: touch.position.y - touch.startPosition.y,
    };
  }

  /**
   * Get the number of active touches.
   */
  static getTouchCount(): number {
    let count = 0;
    for (const touch of this.touches.values()) {
      if (touch.isDown) count++;
    }
    return count;
  }

  private static getTouchByIndex(index: number): TouchState | undefined {
    let i = 0;
    for (const touch of this.touches.values()) {
      if (i === index) return touch;
      i++;
    }
    return undefined;
  }

  // ========== Keyboard Input ==========

  /**
   * Handle key event from platform.
   */
  static onKeyEvent(key: KeyCode, pressed: boolean): void {
    let state = this.keys.get(key);

    if (!state) {
      state = { isDown: false, justPressed: false, justReleased: false };
      this.keys.set(key, state);
    }

    if (pressed && !state.isDown) {
      state.isDown = true;
      state.justPressed = true;
    } else if (!pressed && state.isDown) {
      state.isDown = false;
      state.justReleased = true;
    }
  }

  /**
   * Check if a key is currently down.
   */
  static isKeyDown(key: KeyCode): boolean {
    const state = this.keys.get(key);
    return state !== undefined && state.isDown;
  }

  /**
   * Check if a key was just pressed this frame.
   */
  static isKeyPressed(key: KeyCode): boolean {
    const state = this.keys.get(key);
    return state !== undefined && state.justPressed;
  }

  /**
   * Check if a key was just released this frame.
   */
  static isKeyReleased(key: KeyCode): boolean {
    const state = this.keys.get(key);
    return state !== undefined && state.justReleased;
  }

  // ========== Mouse Input ==========

  /**
   * Handle mouse move event from platform.
   */
  static onMouseMove(x: number, y: number): void {
    this.mousePosition = { x, y };
  }

  /**
   * Handle mouse button event from platform.
   */
  static onMouseButton(button: number, pressed: boolean): void {
    let state = this.mouseButtons.get(button);

    if (!state) {
      state = { isDown: false, justPressed: false, justReleased: false };
      this.mouseButtons.set(button, state);
    }

    if (pressed && !state.isDown) {
      state.isDown = true;
      state.justPressed = true;
    } else if (!pressed && state.isDown) {
      state.isDown = false;
      state.justReleased = true;
    }
  }

  /**
   * Get the current mouse position.
   */
  static getMousePosition(): Vec2 {
    return { ...this.mousePosition };
  }

  /**
   * Check if a mouse button is currently down.
   */
  static isMouseButtonDown(button: number = 0): boolean {
    const state = this.mouseButtons.get(button);
    return state !== undefined && state.isDown;
  }

  /**
   * Check if a mouse button was just pressed.
   */
  static isMouseButtonPressed(button: number = 0): boolean {
    const state = this.mouseButtons.get(button);
    return state !== undefined && state.justPressed;
  }

  /**
   * Check if a mouse button was just released.
   */
  static isMouseButtonReleased(button: number = 0): boolean {
    const state = this.mouseButtons.get(button);
    return state !== undefined && state.justReleased;
  }
}
